async/src/scheduler.rs

```diff
@@ -1,99 +1,135 @@
     util::tracing,
 };
 use cordyceps::mpsc_queue::MpscQueue;
 use core::{future::Future, pin::Pin};
 
-#[derive(Clone, Debug, Default)]
-pub struct Scheduler(Arc<Core>);
+#[cfg(feature = "alloc")]
+pub use arc_scheduler::*;
 
-#[derive(Debug, Default)]
-pub struct StaticScheduler(Core);
-
-#[derive(Debug)]
-struct Core {
-    run_queue: MpscQueue<Header>,
-    // woken: AtomicBool,
-}
+#[cfg(feature = "alloc")]
+mod arc_scheduler {
+    use super::*;
+    use crate::{
+        loom::sync::Arc,
+        task::{allocation::BoxStorage, Task},
+    };
+    use alloc::boxed::Box;
 
-#[derive(Debug)]
-#[non_exhaustive]
-pub struct Tick {
-    pub polled: usize,
-    pub completed: usize,
-    pub has_remaining: bool,
-}
+    #[derive(Clone, Debug, Default)]
+    pub struct Scheduler(Arc<Core>);
 
-/// A trait abstracting over spawning futures.
-pub trait Spawn<F: Future> {
-    /// Spawns `future` as a new task on this executor.
-    fn spawn(&self, future: F);
-}
+    // === impl Scheduler ===
 
-pub(crate) trait Schedule: Sized + Clone {
-    fn schedule(&self, task: TaskRef);
-}
+    impl Scheduler {
+        /// How many tasks are polled per call to `Scheduler::tick`.
+        ///
+        /// Chosen by fair dice roll, guaranteed to be random.
+        pub const DEFAULT_TICK_SIZE: usize = Core::DEFAULT_TICK_SIZE;
 
-// === impl Scheduler ===
+        pub fn new() -> Self {
+            Self::default()
+        }
 
-impl Scheduler {
-    /// How many tasks are polled per call to `Scheduler::tick`.
-    ///
-    /// Chosen by fair dice roll, guaranteed to be random.
-    pub const DEFAULT_TICK_SIZE: usize = Core::DEFAULT_TICK_SIZE;
+        #[inline]
+        pub fn spawn(&self, future: impl Future) {
+            self.schedule(TaskRef::new(self.clone(), future));
+        }
 
-    pub fn new() -> Self {
-        Self::default()
-    }
+        #[inline]
+        pub fn spawn_allocated<F>(&'static self, task: Box<Task<Self, F, BoxStorage>>)
+        where
+            F: Future,
+        {
+            let tr = TaskRef::new_allocated::<Self, F, BoxStorage>(task);
+            self.schedule(tr);
+        }
 
-    #[inline]
-    pub fn spawn(&self, future: impl Future) {
-        Core::spawn_arc(&self.0, future)
+        pub fn tick(&self) -> Tick {
+            self.0.tick_n(Self::DEFAULT_TICK_SIZE)
+        }
     }
 
-    pub fn tick(&self) -> Tick {
-        self.0.tick_n(Self::DEFAULT_TICK_SIZE)
+    impl Schedule for Scheduler {
+        fn schedule(&self, task: TaskRef) {
+            // self.woken.store(true, Ordering::Release);
+            self.0.run_queue.enqueue(task);
+        }
     }
 }
 
-/// A trait abstracting over spawning futures.
-impl<F: Future> Spawn<F> for Scheduler {
-    /// Spawns `future` as a new task on this executor.
-    fn spawn(&self, future: F) {
-        Scheduler::spawn(self, future)
-    }
-}
+pub use static_scheduler::*;
 
-// === impl StaticScheduler ===
+mod static_scheduler {
```

Comment:
By the way, I added these mostly to keep things organize for myself, if you want me to revert this I can.